Persistent storage yuhang liao

ComarPers922 opened the pull request:
Persistent storage using serialization.

ComarPers922 committed:
Persistent Storage

ComarPers922 committed:
User can use the initial parameter to specify the file path for storage.

ComarPers922 committed:
Change the class name from Storage to LocalFile.

ComarPers922 committed:
Improve the performance.

ComarPers922 committed:
Merge branch 'Version-Check-Yuhang-Liao' of github.com:vgomezASC/CodeU-Summer-2017 into Version-Check-Yuhang-Liao

ComarPers922 committed:
Improve the performance.

ComarPers922 committed:
Friendly Output.

aslo commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -120,11 +126,30 @@ public void invoke(List<String> args) {
+    //
+    // Get some infomation from server; it should be version info currently
+    //
+    panel.register("info", new Panel.Command() {
```
Hmm, it's surprising to me that this is showing up in the changes for this PR. This was already merged into master, so it should not appear as being an addition as part of this PR.

I think this means you need to pull the latest changes from master into this branch. Can you run `git pull origin master`, then `git push`? That should update your branch so we only see the lines that have changed as part of the persistent storage feature.

aslo commented on src/codeu/chat/server/LocalFile.java:
```diff
@@ -0,0 +1,166 @@
+            fileInputStream.close();
+            objectInputStream.close();
+        }
+        catch(Exception exception)
```
Generally, it's not good to catch all `Exception`s in a single block like this. It's better to be specific and catch each type of exception that you expect, and react to them individually. 

In this case, the above block of code can throw an IOException and a ClassNotFoundException. If there's an IOException, that can indicate that there was a temporary error with the file system. A ClassNotFoundException on the other hand, indicates that the program is missing the classes it needs to deserialize objects.

I'd recommend using a message that indicates exactly what went wrong, and also just allowing the program to crash (see my comment below for more on that).

```java
try {

// read from file system and deserialize

} catch (IOException e) {
  System.err.println("ERROR: A transient error was encountered while reading from a local file. Please try again.");
  throw new RuntimeException(e);
  
} catch (ClassNotFoundException e) {
  System.err.println("ERROR: There was an error deserializing data. Required classes were not found.");    
  throw new RuntimeException(e);
}
```

aslo commented on src/codeu/chat/server/LocalFile.java:
```diff
@@ -0,0 +1,166 @@
+        finally
+        {
+            //Write temporary information to the instance varibles
+            users = tempUsers;
```
I don't think this `finally` block is a good idea. If the code in the `try` block fails, tempUsers, tempConversationHeaders, and tempMessages will all be null.  If that happens, the program will enter a weird state, where it could potentially overwrite valid users/conversations/messages in the file the next time that it saves. 

In my opinion, you should remove this `finally` block.  If there are exceptions above, you should allow the program to crash rather than assigning these objects to null.

aslo commented on src/codeu/chat/server/LocalFile.java:
```diff
@@ -0,0 +1,166 @@
+        {
+            return;
+        }
+        hasModified = false;
```
You should move this line to the bottom of this method - only mark `hasModified` as false once the save has successfully completed. Currently, if there's an IOException while attempting to save, hasModified will be in an incorrect state.

aslo commented on src/codeu/chat/server/LocalFile.java:
```diff
@@ -0,0 +1,166 @@
+            file.createNewFile();
+        }
+        //Serialization
+        RandomAccessFile randomAccessFile = new RandomAccessFile(file, "rw");
```
Why did you choose to use a `RandomAccessFile` here? I think you can just do 
```java
FileOutputStream fileOutputStream = new FileOutputStream(file);
ObjectOutputStream objectOutputStream = new ObjectOutputStream(fileOutputStream)
```

aslo commented on src/codeu/chat/server/LocalFile.java:
```diff
@@ -0,0 +1,166 @@
+
+        LOG.info("Data saved!");
+
+        fileOutputStream.close();
```
If there's an exception in this method, these lines won't be executed, and these streams won't ever be closed. This is a memory leak. A nice way of avoiding this is to use a ["try-with-resources"](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) block, which is a relatively new Java feature. It will automatically close these streams when you're done with them, even if there's an exception. Here's what it would look like: 

```java
try (FileOutputStream fileOutputStream = new FileOutputStream(randomAccessFile.getFD());
     ObjectOutputStream objectOutputStream = new ObjectOutputStream(fileOutputStream) ) {

       ObjectOutputStream objectOutputStream = new ObjectOutputStream(fileOutputStream);
       objectOutputStream.writeObject(this); 
       LOG.info("Data saved!");
}
```

aslo commented on src/codeu/chat/server/Server.java:
```diff
@@ -193,6 +213,23 @@ public void run() {
+        try
+        {
+          localFile.saveData();
+          timeline.scheduleIn(LOCAL_FILE_REFRESH_MS, this);
```
If the above line throws an exception, this won't be rescheduled. This line should be executed inside a `finally` block to make sure that you always schedule the next refresh.

aslo reviewed (changes requested):
This is looking great! I left you a few comments but I'm very impressed with the progress.

Also, one comment that github doesn't allow me to make inline - you should not commit the `dat.sav` file with the code. It would be a good idea to add it to the .gitignore file.

ComarPers922 committed:
Merge branch 'master' of github.com:vgomezASC/CodeU-Summer-2017 into Persistent-Storage-Yuhang-Liao

ComarPers922 committed:
Fix some bugs.

ComarPers922 committed:
dat.sav will not be pushed

ComarPers922 committed:
Improve the performance.

ComarPers922 committed:
*.sav will not be pushed

ComarPers922 committed:
Delete dat.sav

ComarPers922 committed:
dat.sav will not be pushed

ComarPers922 committed:
Modify the format

aslo commented on src/codeu/chat/server/Controller.java:
```diff
@@ -34,9 +35,38 @@
+      this.newMessage(item.id, item.author, item.conversation, item.content, item.creation);
+    }
+    //Initialization Complete
+    isInitialized = true;
```
Why do you have this variable?  It looks like it will always be true, no matter what, which means you should get rid of it.

aslo commented on src/codeu/chat/server/Controller.java:
```diff
@@ -64,9 +94,21 @@ public Message newMessage(Uuid id, Uuid author, Uuid conversation, String body,
+      message = new Message(id, Uuid.NULL, Uuid.NULL, creationTime, author, body,conversation);
       model.add(message);
-      LOG.info("Message added: %s", message.id);
+      if(localFile != null && isInitialized)
```
You should not allow `localFile` to be null, because if it is null, this class will not be in a working state. You can guarantee that it's never null by deleting the constructor that doesn't include it (the one that looks like `public Controller(Uuid serverId, Model model)`)

aslo commented on src/codeu/chat/server/LocalFile.java:
```diff
@@ -0,0 +1,170 @@
+import codeu.chat.util.Logger;
+import codeu.chat.util.Uuid;
+
+public class LocalFile implements Serializable
```
I don't believe this class needs to implement Serializable, since it's never serialized

aslo commented on src/codeu/chat/server/LocalFile.java:
```diff
@@ -0,0 +1,170 @@
+		    ObjectInputStream objectInputStream = new ObjectInputStream(fileInputStream))
+        {
+            //Read the information to current instance
+		    newClass = (LocalFile)objectInputStream.readObject();
```
This is strange for a class to create new instances of itself by reading them in from an ObjectInputStream like this. Actually, in general it's not recommended to use java object serialization like this at all. Here's an article on this with more on why: https://softwareengineering.stackexchange.com/questions/191269/java-serialization-advantages-and-disadvantages-use-or-avoid

Instead, I'd recommend using the custom serializers that are already available in the (Message, Conversation, and User classes)[https://github.com/vgomezASC/CodeU-Summer-2017/blob/master/src/codeu/chat/common/Message.java#L28]. These serializers provide the same functionality, and you can use them to read/write these objects to and from a FileInputStream and FileOutputStream.

aslo commented on src/codeu/chat/server/LocalFile.java:
```diff
@@ -0,0 +1,170 @@
+            throw new RuntimeException("ERROR:Unable to run the program. Local file cannot be read.");
+        }
+       //Write temporary information to the instance varibles
+       users = tempUsers;
```
You can delete these lines, and remove the use of these `temp` variables now that you are properly handling exceptions in this constructor - just assign `users`, `conversationHeaders`, and `messages` in the `try` block when you deserialize.

aslo commented on src/codeu/chat/server/LocalFile.java:
```diff
@@ -0,0 +1,170 @@
+        try(FileOutputStream fileOutputStream = new FileOutputStream(file);
+        ObjectOutputStream objectOutputStream = new ObjectOutputStream(fileOutputStream))
+        {
+         objectOutputStream.writeObject(this); 
```
See my above comment - rather than using java object serialization, I'd recommend using the custom SERIALIZER classes in the message, conversation, and user objects instead.

ComarPers922 committed:
Implementation of Serializable is not required.

ComarPers922 committed:
Ignore all .sav file.

ComarPers922 committed:
Good-looking

ComarPers922 committed:
1.Implementation of Serializable is not required.
2.Modify the serializer, so it can serialize the conversation data.

ComarPers922 committed:
New file management method.

ComarPers922 committed:
Don't need built-in Serializer anymore.

ComarPers922 committed:
Data manager.

ComarPers922 committed:
Fix some bugs.

ComarPers922 committed:
Fix some bugs.

ComarPers922 committed:
Fix some bugs.

ComarPers922 committed:
Fix some bugs.

aslo commented on src/codeu/chat/server/Controller.java:
```diff
@@ -34,9 +43,79 @@
+  private final File conversationFile;
+  private final File messageFile;
+
+  private boolean isInitialized = false;
```
Don't forget to get rid if `isInitialized`! It's not necessary.

aslo commented on src/codeu/chat/server/Controller.java:
```diff
@@ -34,9 +43,79 @@
+    messageFile = new File(localFile.getPath() + LocalFile.MESSAGE_FILE_NAME);
+    try
+    {
+      FileInputStream userInputStream = new FileInputStream(userFile);
```
This is a good opportuntity to use a try-with-resources statement, because the stream will automatically be cleaned up when you're done with it.

```
try (FileInputStream userInputStream = new FileInputStream(userFile)) {
  // ... etc
} 
```

aslo commented on src/codeu/chat/server/Controller.java:
```diff
@@ -34,9 +43,79 @@
+       }
+      }
+      
+      FileInputStream conversationInputStream = new FileInputStream(conversationFile);
```
You should also use a try-with-resources statement here.

aslo commented on src/codeu/chat/server/Controller.java:
```diff
@@ -34,9 +43,79 @@
+        }
+      }
+
+      FileInputStream messageInputStream = new FileInputStream(messageFile);
```
You should also use a try-with-resources statement here.

aslo commented on src/codeu/chat/server/Controller.java:
```diff
@@ -34,9 +43,79 @@
+    
+    this.localFile = localFile;//The path is assigned by server.
+
+    userFile = new File(localFile.getPath() + LocalFile.USER_FILE_NAME);
```
I'd recommend moving all of this code, that reads from your save files, into the `LocalFile` class. It's usually good to group similar functionality in the same place, and I think it would be good to group all reading and writing to/from files together.

aslo commented on src/codeu/chat/server/LocalFile.java:
```diff
@@ -0,0 +1,299 @@
+
+    //false: The instance is not initialized. All new data will not mark the status as modified.
+    //true: All new data been added to this instance should mark the status as modified.
+    private boolean isInitialized = false;
```
You should delete `isInitialized` here as well. This class should always be initialized if its constructor completes successfully.

aslo commented on src/codeu/chat/server/LocalFile.java:
```diff
@@ -0,0 +1,299 @@
+     */
+    public LinkedHashSet<Message> getCopyOfMessages()
+    {
+        hasMessageModified = false;
```
This seems odd to me. Normally, a "getter" method shouldn't modify any state, and this would be unexpected behavior to anyone working with this class in the future. You should remove this line that sets hasMessageModified = false.

aslo commented on src/codeu/chat/server/LocalFile.java:
```diff
@@ -0,0 +1,299 @@
+     */
+    public LinkedHashSet<ConversationHeader> getCopyOfConversationHeaders()
+    {
+        hasConversationModified = false;
```
This seems odd to me. Normally, a "getter" method shouldn't modify any state, and this would be unexpected behavior to anyone working with this class in the future. You should remove this line that sets hasConversationModified = false.

aslo commented on src/codeu/chat/server/LocalFile.java:
```diff
@@ -0,0 +1,299 @@
+     */
+    public LinkedHashSet<User> getCopyOfUsers()
+    {
+        hasUserModified = false;
```
This seems odd to me. Normally, a "getter" method shouldn't modify any state, and this would be unexpected behavior to anyone working with this class in the future. You should remove this line that sets hasUserModified = false.

aslo commented on src/codeu/chat/common/Message.java:
```diff
@@ -60,6 +61,7 @@ public Message read(InputStream in) throws IOException {
   public Uuid next;
+  public final Uuid conversation;
 
   public Message(Uuid id, Uuid next, Uuid previous, Time creation, Uuid author, String content) {
```
You should delete this constructor.

ComarPers922 committed:
Remove an unused constructor.

ComarPers922 committed:
Merge saving and loading into LocalFile class

ComarPers922 committed:
1.Merge read and write into one class.
2.Modify the names of getters. It still return a new copies of three LinkedHashSet in order to use foreach-block.
